pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -70,126 +76,244 @@ var (
 	}
 )
 
-// Requirements is a decorated alias type for []v1.NodeSelectorRequirements
-type Requirements []v1.NodeSelectorRequirement
-
-func (r Requirements) Zones() sets.String {
-	return r.Requirement(v1.LabelTopologyZone)
-}
-
-func (r Requirements) InstanceTypes() sets.String {
-	return r.Requirement(v1.LabelInstanceTypeStable)
+type Requirements struct {
+	// Requirements are layered with Labels and applied to every node.
+	Requirements []v1.NodeSelectorRequirement `json:"requirements,omitempty"`
+	// Ground truth for record keeping
+	Records map[string][]v1.NodeSelectorRequirement `json:"-"`
+	Allows  map[string]sets.Set                     `json:"-"`
 }
 
-func (r Requirements) Architectures() sets.String {
-	return r.Requirement(v1.LabelArchStable)
+func NewRequirements(requirements ...v1.NodeSelectorRequirement) *Requirements {
+	result := Requirements{
+		Records: map[string][]v1.NodeSelectorRequirement{},
+		Allows:  map[string]sets.Set{},
+	}
+	return result.Add(requirements...)
 }
 
-func (r Requirements) OperatingSystems() sets.String {
-	return r.Requirement(v1.LabelOSStable)
+func (r *Requirements) Add(requirements ...v1.NodeSelectorRequirement) *Requirements {
+	if r == nil {
+		return NewRequirements(requirements...)
+	}
+	rd := r.DeepCopy()
+	for _, requirement := range requirements {
+		// Normalizing the key
+		if newKey, ok := NormalizedLabels[requirement.Key]; ok {
+			requirement.Key = newKey
+		}
+		//rd.Requirements = append(rd.Requirements, requirement)
+		rd.Records[requirement.Key] = append(rd.Records[requirement.Key], requirement)
+		switch requirement.Operator {
+		case v1.NodeSelectorOpIn:
+			rd.Allows[requirement.Key] = rd.Allow(requirement.Key).Intersection(sets.NewSet(requirement.Values...))
+		case v1.NodeSelectorOpNotIn:
+			rd.Allows[requirement.Key] = rd.Allow(requirement.Key).Intersection(sets.NewComplementSet(requirement.Values...))
+		}
+	}
+	return rd
 }
 
-func (r Requirements) CapacityTypes() sets.String {
-	return r.Requirement(LabelCapacityType)
+func (r *Requirements) MarshalJSON() ([]byte, error) {
+	var result []v1.NodeSelectorRequirement
+	for _, requirements := range r.Records {
```

Comment:
Can't we do this via r.Requirements?

Reply:
currently r.Requirements is just a nil pointer. I never populate it so it contains no data. This way we save some memory.